### PR TITLE
Add cni bash script

### PR DESCRIPTION
**Pull request by rhrazdil**
Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

This PR adds a cni bash script taht configures nftable rules to drop ingress/egress traffic that is not explicitly allowed.

example net-attach-def for the plugin:
```yaml
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
 name: a-policy-network
spec:
 config: '{
   "cniVersion":"0.3.1",
   "name":"nad-bridge-cidr",
   "plugins":[
      {
         "type":"cnv-bridge",
         "bridge":"br1"
      },
      {
         "type":"cidr-filtering-cni",
         "name":"p1",
         "ingress":{
            "subnets":[
               {
                  "subnet":{
                     "cidr":"192.168.66.0/24",
                     "except":[
                        "192.168.66.102",
                        "192.168.66.103",
                        "192.168.66.104"
                     ]
                  }
               },
               {
                  "subnet":{
                     "cidr":"1.1.1.0/24",
                     "except":[
                        "1.1.1.2"
                     ]
                  }
               }
            ],
            "ports":[
               {
                  "protocol":"TCP",
                  "port":"1-65535"
               },
               {
                  "protocol":"udp",
                  "port":"1-65535"
               }
            ]
         },
         "egress":{
            "subnets":[
               {
                  "subnet":{
                     "cidr":"192.168.66.0/24",
                     "except":[
                        "192.168.66.102"
                     ]
                  }
               },
               {
                  "subnet":{
                     "cidr":"255.255.255.255/32",
                     "except":[]
                  }
               },
               {
                  "subnet":{
                     "cidr":"1.1.1.0/24"
                  }
               }
            ],
            "ports":[
               {
                  "protocol":"TCP",
                  "port":"1-65535"
               },
               {
                  "protocol":"udp",
                  "port":"1-65535"
               }
            ]
         }
      }
   ]
   }'
```

And the nftable ruleset that is created based on the above configuration:

```
table bridge filter {
	chain postrouting {
		type filter hook postrouting priority 0; policy accept;
		ether type arp counter packets 5 bytes 140 accept
		oifname "net1" counter packets 26 bytes 3072 jump m-egress
	}

	chain m-egress {
		counter packets 26 bytes 3072 meta mark set meta mark & 0xfffcffff
		oifname "net1" counter packets 26 bytes 3072 jump m-03a8c-egress
		meta mark & 0x00030000 == 0x00030000 counter packets 6 bytes 1456 return
		counter packets 20 bytes 1616 drop
	}

	chain m-03a8c-egress {
		counter packets 26 bytes 3072 jump m-03a8c-egress-subnets
		counter packets 26 bytes 3072 jump m-03a8c-egress-ports
	}

	chain m-03a8c-egress-subnets {
		oifname "net1" ip daddr 192.168.66.102 counter packets 0 bytes 0 drop
		oifname "net1" ip daddr 192.168.66.0/24 counter packets 2 bytes 134 meta mark set meta mark | 0x00020000
		oifname "net1" ip daddr 255.255.255.255 counter packets 4 bytes 1322 meta mark set meta mark | 0x00020000
		oifname "net1" ip daddr 1.1.1.0/24 counter packets 0 bytes 0 meta mark set meta mark | 0x00020000
	}

	chain m-03a8c-egress-ports {
		oifname "net1" tcp dport 1-65535 counter packets 0 bytes 0 meta mark set meta mark | 0x00010000
		oifname "net1" udp dport 1-65535 counter packets 10 bytes 1760 meta mark set meta mark | 0x00010000
	}
}
table netdev filter {
	chain ingress {
		type filter hook ingress device "net1" priority filter; policy accept;
		ether type arp counter packets 4 bytes 112 accept
		iifname "net1" counter packets 12 bytes 1982 jump m-ingress
	}

	chain m-ingress {
		counter packets 12 bytes 1982 meta mark set meta mark & 0xfffcffff
		iifname "net1" counter packets 12 bytes 1982 jump m-03a8c-ingress
		meta mark & 0x00030000 == 0x00030000 counter packets 5 bytes 1294 return
		counter packets 2 bytes 232 drop
	}

	chain m-03a8c-ingress {
		counter packets 12 bytes 1982 jump m-03a8c-ingress-subnets
		counter packets 12 bytes 1982 jump m-03a8c-ingress-ports
	}

	chain m-03a8c-ingress-subnets {
		iifname "net1" ip saddr 192.168.66.102 counter packets 0 bytes 0 drop
		iifname "net1" ip saddr 192.168.66.103 counter packets 0 bytes 0 drop
		iifname "net1" ip saddr 192.168.66.104 counter packets 0 bytes 0 drop
		iifname "net1" ip saddr 192.168.66.0/24 counter packets 5 bytes 1294 meta mark set meta mark | 0x00020000
		iifname "net1" ip saddr 1.1.1.2 counter packets 0 bytes 0 drop
		iifname "net1" ip saddr 1.1.1.0/24 counter packets 0 bytes 0 meta mark set meta mark | 0x00020000
	}

	chain m-03a8c-ingress-ports {
		iifname "net1" tcp dport 1-65535 counter packets 0 bytes 0 meta mark set meta mark | 0x00010000
		iifname "net1" udp dport 1-65535 counter packets 5 bytes 1294 meta mark set meta mark | 0x00010000
	}
}

```

Notes:
1)
The plugin uses `ingress` and `postrouting` chains to filter the traffic.
The reason for using `ingress` instead of `prerouting` is that `prerouting` chain for some reason doesn't
see any traffic, when used with KubeVirt.
This comes with an issue, ingress hook doesn't have yet reconstructed packets, so filtering at L4 doesn't work for fragmented traffic. 
Since we'll need to support L4 filtering too (ports), this is a temporary workaround.

### CODE REVIEW

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni]: shouldn't this be executed in the pod's netns ? ...

Comment (maiqueb) [reply cidr-filtering-cni]: once this is fixed I am able to start the VM.

**Review by maiqueb — COMMENTED** [review]
We need to be able to properly report errors. I had to debug this and had no choice but to be editing the script manually and adding `echo` statements.

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni]: hm, not exactly sure - a log file could probably come in handy. 

Feel free to remove it, then re-add a better "version" in a follow-up PR.

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni]: egress rules must match on the `oifname`, and as part of the post routing hook.

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni]: you'll need to create 2 base chains - one for ingress, listening to the prerouting hook, using iifname - another for egress - listening to the postrouting hook, using oifname.

**Comment by rhrazdil** [comment]
Updated the cni script to create sub chains: [force-push](https://github.com/rhrazdil/cidr-filtering-cni/compare/961119ca6ea94ba8a8127d7c6df419faae234739..98a05e0209953075812b7c87edba8439f49dbfb1)

- Ingress traffic is filtered at prerouting
- Egress traffic si filtered at postrouting

Ports are not yet handled, and simply all ports are allowed.

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni]: iiuc, this means you'll only accept a single policy per net-attach-def. 

Is that intentional ? Would it make sense for a user to reference multiple policies ? The idea being you could re-use policies across users.

Comment (rhrazdil) [reply cidr-filtering-cni]: Ah thanks, I didn't realize that use case. This makes sense.

So in the end, the idea is to reference MultiNetworkPolicies in net-attach-def and the CNI applies the rules accordingly to the current state of the policies, does it sound right?

Comment (maiqueb) [reply cidr-filtering-cni]: That's the thing: no. 

`MultiNetworkPolicies` refer to net-attach-defs via a this [`policy-for`](https://github.com/k8snetworkplumbingwg/multi-networkpolicy#policy-for-annotation) annotation - so it is the other way around.

The intended implementation is - imo - clearly meant for a controller based approach. Using this API, you'd have to filter all the relevant policies that point to the NAD being used. It can be done, but it will take more time.

Lasty, one thing that I would also like to point out is I am unsure how much sense it makes to follow the multi-network-policy API at all - we're only allowing a subset of it (no namespace nor label based policies ...). 

The advantage of using the multi-network policy is quite obvious: users wanting to migrate to full-fledged solutions of it would be able to do so seamlessly. The disadvantage would be they could be confused / surprised at the fact only a subset of it is implemented.

File: README.md
Comment (maiqueb) [inline README.md]: I think the API is wrong - the port range should be scoped to the ipBlock.

I assume you'll be associating the ipBlocks to the port ranges based on index - no point doing that if you just have the port definition within the IP block.

EDIT: I have re-reviewed the API of the k8s network policies, and this seems to follow their mental model. I stand corrected. Disregard this comment please.

File: README.md
Comment (maiqueb) [inline README.md]: It should also feature ranges, for instance:

Allowed network CIDR: 10.10.0.0/16
Excluded:
- 10.10.1.0/24
- 10.10.2.0/24
- 10.10.3.14
- 10.10.4.18

Just an example of what is allowed via the multi-net policies. Technically it is no different to block an IP than a range.

File: README.md
Comment (maiqueb) [inline README.md]: nit: Remove these newlines

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni:4]: you're using so many variables below that I would consider using `set -u`, to explicitly error on unbounded variables with an error like ` "XYZ: unbound variable"` . 

Not sure if using -x would also be helpful.

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni]: I think we may want to have this overridden, using this default.

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni:15]: we should be using the prerouting hook.

Using the ingress hooks features limitations:
- only available from kernel 4.2 onwards
- the packets were not yet reassembled - meaning it will only work for unfragmented packets, or the first fragmented packet.

I'm unsure if we should block this PR until you manage to sort out why you're not being able to match on the pre-routing hook (other than ARP frames), or should accept it and work on fixing it on top.

File: README.md
Comment (maiqueb) [inline README.md:1]: I miss a `Requirements` section in the readme.

AFAIU, we rely on:
- jq
- nft
- sha1sum

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni]: `fn` maybe ? That really does scream "function" and I was honestly thinking "file" when I saw this stray `f` . 

Maybe it's just me, not a must.

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni:49]: I'm afraid this is too broad ... but since we don't want to have many dependencies on the host, and I'm mostly clueless at bash, I don't have a better suggestion.

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni]: Does it make sense to use a different algorithm here ? A symmetric one, so you can also go from `policy_id` to `policy_name` .

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni:137]: is lowercasing the protocol really required ? (honest question, I got no clue).

File: cidr-filtering-cni
Comment (oshoval) [inline cidr-filtering-cni]: Do we want to add it as in a main function maybe ?

Unless we can't because it is a CNI
btw there are no prints to stdout beside the final answer right ? (CNI doesnt allow it)

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni:149]: so, to make sure I understand: the ports would apply for **all** the IP blocks, right ?

File: cidr-filtering-cni
Comment (oshoval) [inline cidr-filtering-cni]: maybe add CHECK stub, and exit 1 for the rest ?
unless there are more commands other than those 4

File: cidr-filtering-cni
Comment (oshoval) [inline cidr-filtering-cni]: not sure we need local word again ?

File: cidr-filtering-cni
Comment (oshoval) [inline cidr-filtering-cni:177]: please make sure that the fact you have `|| exitWithError` doesn't make this function disregard -e all over
there is a gotcha in bash about it IIRC

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni]: here we need to make sure to also write to the bridge table ( I'm aware there's the bug where only ARP frames are matched ..).

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni]: Would it make sense to only create the *relevant* pipelines ? I mean only create the ingress pipeline for ingress policies, and only create the egress pipeline for egress pipeline.

If a policy features both, then you should of course create both chains.

File: cidr-filtering-cni
Comment (oshoval) [inline cidr-filtering-cni]: but there aren't actually usages of more than 3 params on `for_json_array` calls right ?
so maybe we dont need this extra support for $@ ?
(it is nice, but make it little heavier)

**Review by oshoval — COMMENTED** [review]
Nice thanks
few nits about bash

File: cidr-filtering-cni
Comment (maiqueb) [inline cidr-filtering-cni]: I think we can get away with supporting versions < 0.4.0. Smart move ... If we used 0.4.0 you'd have to also implement cmd check.

**Review by maiqueb — CHANGES_REQUESTED** [review]
What worries me more is hooking in the netdev table using `ingress`.

I think we need to address this ASAP.

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni:137]: It is, I've tried

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni:149]: Yes, that is correct

File: README.md
Comment (rhrazdil) [reply README.md]: Right, of course. I think it should work out of the box, I need to check it and fix the docs

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni]: the name length of a nft chain is limited to 32 chars, so even with symmetric algorithm, I would need to strip the result.
So I wouldn't be able to get to the original string anyway.

Comment (rhrazdil) [reply cidr-filtering-cni]: encryption algorithm would also likely yield unprintable characters, that would need to be dealt with.
If we want to make it easier for us, it would be maybe better to not do this at all and just strip the original name to the desired length. If two or more configurations happen to have the same name, it would still work we would just end up with all the rules in one chain.

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni]: > Do we want to add it as in a main function maybe ?

I'm not sure what you mean. 

> btw there are no prints to stdout beside the final answer right ? (CNI doesnt allow it)

that is correct, the only the result should be printed to stdout

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni:177]: I tested the behaviour and it seems to work as desired. Don't you mix it up with a pipe?

Because for example `false | echo hello ` succeeds with $? == 0. That is why I  `set -o pipefail` to carry failures forward when in piped commands

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni]: Good point. I'll exit with unrecognized CNI command.
We don't need to deal with CHECK, since that is introduced in version 0.4.0 (and we don't need to support that IIUC)

File: cidr-filtering-cni
Comment (oshoval) [reply cidr-filtering-cni]: > > Do we want to add it as in a main function maybe ?
> 
> I'm not sure what you mean.

```
function main() {
}

main "$@"
```

File: cidr-filtering-cni
Comment (oshoval) [reply cidr-filtering-cni:177]: not mixing it with a pipe, if a function has `|| return true` on a command, it will disregard the set -e
(i might not remember the exact combination, but it wasn't a pipe)
anyhow if it works good then all good

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni]: I see, we can do that

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni]: I think for start we could create
m-"${policy_id}"-"${direction}"-[subnets,ports]
chains only when we know there's a `subnets` or `ports` key.

That way we wouldn't jump to empty chains for matching subnets/ports.

Normally, I think we can assume there would be at least some ports and some IPs allowed.

If you don't mind, I would rather address this in the followup PR that reads configurations from configmaps, as that would conflict with this change a fair bit.

**Comment by oshoval** [comment]
`jq` all over, maybe either mention its needed on the readme, or verify it exists in the script to exit error nicely ?
EDIT - ah sorry saw now it is already commented about

File: README.md
Comment (rhrazdil) [reply README.md:1]: Done

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni:4]: Added -u.

I'll see about -x, I suppse I would need to redirect stdout to log file, until we actually write a result.

Comment (oshoval) [reply cidr-filtering-cni:4]: I think better to not have -x
whatever needed can be added with a log command to a log file / stderr

Comment (maiqueb) [reply cidr-filtering-cni:4]: I'm sold - forget about `-x` . 

Let the logger log.

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni]: Yep, removed

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni]: Indeed, it's not used. Removed

File: cidr-filtering-cni
Comment (oshoval) [inline cidr-filtering-cni]: btw why not one liner?
`set -euo pipefail`

File: manifests/daemonset.yaml
Comment (oshoval) [inline manifests/daemonset.yaml]: is this updated automatically ? 
maybe worth to have a script that does it + CI job that checks it was updated ?
easy to forget otherwise

File: cidr-filtering-cni
Comment (oshoval) [inline cidr-filtering-cni]: maybe print the cmd itself as well ?

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni]: Done

File: manifests/daemonset.yaml
Comment (maiqueb) [reply manifests/daemonset.yaml]: While I agree this can go on an automation follow-up.

Comment (rhrazdil) [reply manifests/daemonset.yaml]: Yep, we'll need to setup CI checking this, for now it's just a `make generate` target

Comment (oshoval) [reply manifests/daemonset.yaml]: np, but just before its merged, lets have the latest base64 if possible

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni]: Yep, Done

File: cidr-filtering-cni
Comment (rhrazdil) [reply cidr-filtering-cni]: Done

**Comment by rhrazdil** [comment]
To sum up for a followup PR:
- Fix prerouting for ingress
- Don't create chains that are empty
- Setup CI automation

**Pull request merged**
